client/state/imports/actions.js

```diff
@@ -0,0 +1,227 @@
+/**
+ * Internal dependencies
+ */
+import wpLib from 'calypso/lib/wp';
+import {
+	IMPORTS_AUTHORS_SET_MAPPING,
+	IMPORTS_AUTHORS_START_MAPPING,
+	IMPORTS_IMPORT_CANCEL,
+	IMPORTS_IMPORT_LOCK,
+	IMPORTS_IMPORT_RECEIVE,
+	IMPORTS_IMPORT_RESET,
+	IMPORTS_IMPORT_START,
+	IMPORTS_IMPORT_UNLOCK,
+	IMPORTS_START_IMPORTING,
+	IMPORTS_UPLOAD_FAILED,
+	IMPORTS_UPLOAD_COMPLETED,
+	IMPORTS_UPLOAD_SET_PROGRESS,
+	IMPORTS_UPLOAD_START,
+} from 'calypso/state/action-types';
+import { appStates } from './constants';
+import { isImporterLocked } from './selectors';
+import { fromApi, toApi } from './api';
+
+/**
+ * Redux dependencies
+ */
+import 'calypso/state/imports/init';
+
+const wpcom = wpLib.undocumented();
+
+const ID_GENERATOR_PREFIX = 'local-generated-id-';
+
+/*
+ * The following `order` functions prepare objects that can be
+ * sent to the API to accomplish a specific purpose. Instead of
+ * actually calling the API, however, they return the _order_,
+ * or request object, so that the calling function can send it
+ * to the API.
+ */
+
+// Creates a request object to cancel an importer
+const createCancelOrder = ( siteId, importerId ) =>
+	toApi( { importerId, importerState: appStates.CANCEL_PENDING, site: { ID: siteId } } );
+
+// Creates a request to expire an importer session
+const createExpiryOrder = ( siteId, importerId ) =>
+	toApi( { importerId, importerState: appStates.EXPIRE_PENDING, site: { ID: siteId } } );
+
+// Creates a request to clear all import sessions
+const createClearOrder = ( siteId, importerId ) =>
+	toApi( { importerId, importerState: appStates.IMPORT_CLEAR, site: { ID: siteId } } );
+
+// Creates a request object to start performing the actual import
+const createImportOrder = ( importerStatus ) =>
+	toApi( {
+		...importerStatus,
+		importerState: appStates.IMPORTING,
+	} );
+
+export const lockImport = ( importerId ) => ( {
+	type: IMPORTS_IMPORT_LOCK,
+	importerId,
+} );
+
+export const unlockImport = ( importerId ) => ( {
+	type: IMPORTS_IMPORT_UNLOCK,
+	importerId,
+} );
+
+export const receiveImporterStatus = ( importerStatus ) => ( dispatch, getState ) => {
+	const isLocked = isImporterLocked( getState(), importerStatus.importId );
+	dispatch( { type: IMPORTS_IMPORT_RECEIVE, importerStatus, isLocked } );
+};
+
+export const cancelImport = ( siteId, importerId ) => async ( dispatch ) => {
+	dispatch( lockImport( importerId ) );
+
+	const cancelImportAction = {
+		type: IMPORTS_IMPORT_CANCEL,
+		importerId,
+		siteId,
+	};
+	dispatch( cancelImportAction );
+
+	// Bail if this is merely a local importer object because
+	// there is nothing on the server-side to cancel
+	if ( importerId.startsWith( ID_GENERATOR_PREFIX ) ) {
+		return;
+	}
+
+	const data = await wpcom.updateImporter( siteId, createCancelOrder( siteId, importerId ) );
+	dispatch( receiveImporterStatus( data ) );
+};
+
+export const fetchState = ( siteId ) => async ( dispatch ) => {
```

Comment:
I'd very much prefer having a more specific name for this thunk - `fetchImporterState` perhaps?

Reply:
Good idea, renamed to `fetchImporterState`.